GeosCore/diagnostics_mod.F90

```diff
@@ -818,14 +809,115 @@ SUBROUTINE Set_SpcConc_Diags_VVDry( Input_Opt,  State_Chm, State_Diag,     &
 
-    ! Error handling
-    IF ( RC /= GC_SUCCESS ) THEN
-       ErrMsg = 'Error converting species units for archiving diagnostics #2'
+  END SUBROUTINE Set_SpcConc_Diags_VVDry
+!EOC
+!------------------------------------------------------------------------------
+!                  GEOS-Chem Global Chemical Transport Model                  !
+!------------------------------------------------------------------------------
+!BOP
+!
+! !IROUTINE: Set_SpcConc_Diags_MND
+!
+! !DESCRIPTION: Subroutine Set_SpcConc\_Diags\_MND sets several species
+!  concentration diagnostic arrays stored in State_Diag to the instantaneous
+!  State_Chm%Species values (in units of "v/v, dry air").
+!\\
+!\\
+! !INTERFACE:
+!
+  SUBROUTINE Set_SpcConc_Diags_MND  ( Input_Opt,  State_Chm, State_Diag,     &
+                                      State_Grid, State_Met, RC            )
+!
+! !USES:
+!
+    USE Input_Opt_Mod,  ONLY : OptInput
+    USE State_Met_Mod,  ONLY : MetState
+    USE State_Chm_Mod,  ONLY : ChmState
+    USE State_Diag_Mod, ONLY : DgnMap
+    USE State_Diag_Mod, ONLY : DgnState
+    USE State_Grid_Mod, ONLY : GrdState
+    USE UnitConv_Mod,   ONLY : Convert_Spc_Units
+!
+! !INPUT PARAMETERS:
+!
+    TYPE(OptInput),   INTENT(IN)    :: Input_Opt    ! Input Options object
+    TYPE(GrdState),   INTENT(IN)    :: State_Grid   ! Grid State object
+    TYPE(MetState),   INTENT(IN)    :: State_Met    ! Meteorology State obj
+!
+! !INPUT/OUTPUT PARAMETERS:
+!
+    TYPE(ChmState),   INTENT(INOUT) :: State_Chm    ! Chemistry State object
+    TYPE(DgnState),   INTENT(INOUT) :: State_Diag   ! Diagnsotics State object
+!
+! !OUTPUT PARAMETERS:
+!
+    INTEGER,          INTENT(OUT)   :: RC           ! Success or failure?
+!
+! !REVISION HISTORY:
+!  21 Nov 2021 - H.P. Lin    - Initial version
+!  See https://github.com/geoschem/geos-chem for complete history
+!EOP
+!------------------------------------------------------------------------------
+!BOC
+!
+! !LOCAL VARIABLES:
+!
+    ! Scalars
+    INTEGER               :: N, S
+
+    ! Strings
+    CHARACTER(LEN=255)    :: ErrMsg, ThisLoc, OrigUnit
+
+    ! Objects
+    TYPE(DgnMap), POINTER :: mapData
+
+    !====================================================================
+    ! Set_SpcConc_Diags_MND begins here!
+    !====================================================================
+
+    ! Assume success
+    RC      =  GC_SUCCESS
+    ThisLoc = &
+         ' -> at Set_SpcConc_Diags_MND (in GeosCore/diagnostics_mod.F90)'
+
+    ! Verify that incoming State_Chm%Species units are kg/kg dry air.
+    IF ( TRIM( State_Chm%Spc_Units ) /= 'kg/kg dry' ) THEN
+       ErrMsg = 'Incorrect species units in Set_SpcConc_Diags_MND!'
        CALL GC_Error( ErrMsg, RC, ThisLoc )
        RETURN
     ENDIF
 
-  END SUBROUTINE Set_SpcConc_Diags_VVDry
+    ! Convert to molec/cm3
+    CALL Convert_Spc_Units( Input_Opt, State_Chm, State_Grid, State_Met, &
+                            'molec/cm3', RC, OrigUnit=OrigUnit )
+
```

Comment:
The above call, along with the later one to convert back, will slightly change the values of species concentrations. I previously went through and removed all instances of using convert_spc_conc for diagnostics to avoid unnecessary changes. This should be deleted and replaced by inline conversation when setting the state_diag arrays.

Reply:
Thanks for pointing that out! I've pushed these changes to commit [1666335b](https://github.com/geoschem/geos-chem/commit/1666335bda77e786388cc19f214b4d203cbc3117).